src/consensus/aft/test/driver.h

```diff
@@ -4,25 +4,60 @@
 
 #include "consensus/aft/raft.h"
 #include "ds/logger.h"
+#include "logging_stub.h"
 
 #include <chrono>
+#include <random>
 #include <set>
 #include <sstream>
 #include <string>
 #include <unordered_map>
 #include <unordered_set>
 
-#define STUB_LOG 1
-#include "logging_stub.h"
+#define RAFT_DRIVER_OUT std::cout << "<RaftDriver>"
+
+std::string stringify(const std::vector<uint8_t>& v, size_t max_size = 15ul)
+{
+  auto size = std::min(v.size(), max_size);
+  return fmt::format(
+    "[{} bytes] {}", v.size(), std::string(v.begin(), v.begin() + size));
+}
+
+struct LedgerStubProxy_WithLogging : public aft::LedgerStubProxy
+{
+  using LedgerStubProxy::LedgerStubProxy;
+
+  void put_entry(
+    const std::vector<uint8_t>& data,
+    bool globally_committable,
+    bool force_chunk) override
+  {
+    RAFT_DRIVER_OUT << "  Node" << _id << "->>Node" << _id
+                    << ": ledger put s: " << stringify(data) << std::endl;
+    aft::LedgerStubProxy::put_entry(data, globally_committable, force_chunk);
+  }
+
+  void truncate(aft::Index idx) override
+  {
+    RAFT_DRIVER_OUT << "  Node" << _id << "->>Node" << _id
+                    << ": truncate i: " << idx << std::endl;
+    aft::LedgerStubProxy::truncate(idx);
```

Comment:
Minor: Moving these behaviour changes (additional logging) to "override the virtual method" rather than "define the `STUB_LOG` macro", so that they're localised to these driver files that use them.